store/postgres/src/relational/index.rs

```diff
@@ -558,6 +725,118 @@ impl CreateIndex {
     }
 }
 
+#[derive(Debug)]
+pub struct IndexList {
+    indexes: HashMap<String, Vec<CreateIndex>>,
+}
+
+impl IndexList {
+    pub fn load(
+        conn: &mut PgConnection,
+        site: Arc<Site>,
+        store: DeploymentStore,
+    ) -> Result<IndexList, StoreError> {
+        let mut list = IndexList {
+            indexes: HashMap::new(),
+        };
+        let schema_name = site.namespace.clone();
+        let layout = store.layout(conn, site)?;
+        for (_, table) in &layout.tables {
+            let table_name = table.name.as_str();
+            let indexes = catalog::indexes_for_table(conn, schema_name.as_str(), table_name)
+                .map_err(StoreError::from)?;
+            let collect: Vec<CreateIndex> = indexes.into_iter().map(CreateIndex::parse).collect();
+            list.indexes.insert(table_name.to_string(), collect);
+        }
+        Ok(list)
+    }
+
+    pub fn indexes_for_table(
+        &self,
+        namespace: &Namespace,
+        table_name: &String,
+        dest_table: &Table,
+        postponed: bool,
+        concurent_if_not_exist: bool,
+    ) -> Vec<(Option<String>, String)> {
+        let mut arr = vec![];
+        if let Some(vec) = self.indexes.get(table_name) {
+            for ci in vec {
+                if ci.fields_exist_in_dest(dest_table)
+                    && !ci.is_default_non_attr_index()
+                    && !(ci.is_id() && dest_table.immutable)
+                    && postponed == ci.to_postpone()
+                {
+                    if let Ok(sql) = ci
+                        .with_nsp(namespace.to_string())
+                        .to_sql(concurent_if_not_exist, concurent_if_not_exist)
+                    {
+                        arr.push((ci.name(), sql))
+                    }
+                }
+            }
+        }
+        arr
+    }
+    pub fn recreate_invalid_indexes(
+        &self,
+        conn: &mut PgConnection,
+        layout: &Layout,
+    ) -> Result<(), StoreError> {
+        #[derive(QueryableByName, Debug)]
+        struct IndexInfo {
+            #[diesel(sql_type = Bool)]
+            isvalid: bool,
+            #[diesel(sql_type = Bool)]
+            isready: bool,
+        }
+
+        let namespace = &layout.catalog.site.namespace;
+        for table in layout.tables.values() {
+            for (ind_name, create_query) in
+                self.indexes_for_table(namespace, &table.name.to_string(), table, true, true)
+            {
+                if let Some(index_name) = ind_name {
+                    let table_name = table.name.clone();
+                    let query = r#"
+                        SELECT  x.indisvalid           AS isvalid,
+                                x.indisready           AS isready
```

Comment:
@lutter do you think that check for `indisready` from `pg_index` table is needed, or should I just check the `indisvalid` only? I decided to ignore the `indislive` here...

Reply:
Looking at the [docs](https://www.postgresql.org/docs/current/catalog-pg-index.html) I think checking `indisvalid` is enough, though I don't think we should ever see one where `indisready` ir `indislive` are `false`

Reply:
Looking again the the docs it seems to me that `inisvalid` is the most strict criteria of the 3. So it feels that when it's true the other two must also be true. So I'll keep that one only.